File: aulas/aula_ampop_1/main.js

ex_AmpOp('reset');
ex_SG('reset');
ex_DAC('reset');
ex_DAC_refresh_input();

function ex_AmpOp(action) {
  const prefixo = ex_AmpOp.name;

  if (action == "reset") {
    reset(prefixo);
  }

  const VA = parseFloat(document.getElementById(`${prefixo}-VA`).value);
  const VB = parseFloat(document.getElementById(`${prefixo}-VB`).value);
  const MaisVcc = parseFloat(
    document.getElementById(`${prefixo}-MaisVCC`).value
  );
  const MenosVcc = parseFloat(
    document.getElementById(`${prefixo}-MenosVCC`).value
  );

  const RS = 25e3;
  const RF = 100e3;

  const Vo = VB - ((VA - VB) / RS) * RF;
  const Vo_str = Vo.toFixed(2);

  const equations = {
    "eq-1": `\\(V_{n}=V_{p}=AAA\\)`,
    "eq-2": `\\(i_{n}=i_{p}=0\\)`,
    "eq-3": `\\(V_{p}=V_{B}\\)`,
    "eq-4": `\\(V_{n}=V_{p}=V_{B}\\)`,
    "eq-5": `\\(i_{S}=\\dfrac{V_{A}-V_{n}}{R_{s}}=\\dfrac{V_{A}-V_{B}}{R_{S}}\\)`,
    "eq-6": `\\(i_{F}=\\dfrac{V_{n}-V_{o}}{R_{F}}=\\dfrac{V_{B}-V_{o}}{R_{F}}\\)`,
    "eq-7": `\\(i_{S}=i_{F}+i_{n}\\)`,
    "eq-8": `\\(i_{S}=i_{F}+0\\)`,
    "eq-9": `\\(\\dfrac{V_{A}-V_{B}}{R_{S}}=\\dfrac{V_{B}-V_{o}}{R_{F}}\\)`,
    "eq-10": `\\(R_{F} \\cdot \\left(\\dfrac{V_{A}-V_{B}}{R_{S}}\\right)-V_{B}=-V_{o}\\)`,
    "eq-11": `\\(V_{o}=V_{B}-R_{F} \\cdot \\left(\\dfrac{V_{A}-V_{B}}{R_{S}}\\right)\\)`,
    "eq-12": `\\(V_{o}=${VB}-100 \\cdot 10^{3} \\cdot \\left(\\dfrac{${VA}-${VB}}{25 \\cdot 10^{3}}\\right)=${Vo_str}V\\)`,
    "eq-13": `\\(\\begin{cases} as=\\dfrac{1}{3} \\\\[0.1pt]  \\\\ as=\\dfrac{1}{3} \\end{cases}\\)`,
  };

  for (const key in equations) {
    const elementId = `${prefixo}-${key}`;
    const element = document.getElementById(elementId);
    if (element) {
      element.innerHTML = equations[key];
      if (window.MathJax && MathJax.typesetPromise) {
        MathJax.typesetPromise([element]);
      }
    }
  }

  const data_parameters = {
    VA: VA.toFixed(2),
    VB: VB.toFixed(2),
    MaisVcc: MaisVcc.toFixed(2),
    MenosVcc: MenosVcc.toFixed(2),
  };

  const data_config = {
    components: ["V_{A}", "V_{B}", "+V_{cc}", "-V_{cc}"],
    units: ["V", "V", "V", "V"],
  };

  // Envie como JSON string
  const data_parameters_json = JSON.stringify(data_parameters);
  const data_config_json = JSON.stringify(data_config);

  if (typeof init_generic === "function") {
    init_generic(prefixo, data_parameters_json, data_config_json);
  }
}



function ex_SG(action) {
  const prefixo = ex_SG.name;

  if (action == "reset") {
    reset(prefixo);
  }

  const VS = parseFloat(document.getElementById(`${prefixo}-VS`).value);
  const RP = parseFloat(document.getElementById(`${prefixo}-RP`).value);
  const RA = parseFloat(document.getElementById(`${prefixo}-RA`).value);
  const RB = parseFloat(document.getElementById(`${prefixo}-RB`).value);
  const DeltaR = parseFloat(document.getElementById(`${prefixo}-DeltaR`).value);

  const RSGPos = RP + DeltaR;
  const RSGNeg = RP - DeltaR;


  let V2 = VS/2;

  let V1_Sup = VS*(RSGPos)/(RSGPos + RP);
  let V1_Inf = VS*(RSGNeg)/(RSGNeg + RP);
  
  let Vo_Min = (RB/RA)*(V2-V1_Sup);
  let Vo_Max = (RB/RA)*(V2-V1_Inf);
    
  let Vo_Min_Apx = -VS*(RB/RA)*(1/(4*RP))*DeltaR;
  let Vo_Max_Apx = VS*(RB/RA)*(1/(4*RP))*DeltaR;
  let ganho = RB/RA;

  V2 = V2.toFixed(2);

  V1_Sup = V1_Sup.toFixed(2);
  V1_Inf = V1_Inf.toFixed(2);
    
  Vo_Min = Vo_Min.toFixed(3);
  Vo_Max = Vo_Max.toFixed(3);
      
  Vo_Min_Apx = Vo_Min_Apx.toFixed(2);
  Vo_Max_Apx = Vo_Max_Apx.toFixed(2);
  
  ganho = ganho.toFixed(2);  

    
  equations = {
    "eq-1": `\\(V_{2}=V_{S} \\cdot \\left(\\dfrac{R_{P}}{R_{P}+R_{P}}\\right)=\\dfrac{V_{S}}{2}=${V2} V\\)`,
    "eq-2": `\\(V'_{1}=V_{S} \\cdot \\left(\\dfrac{R_{P}+\\Delta R}{R_{P}+\\Delta R+R_{P}}\\right)\\)`,
    "eq-3": `\\(V'_{1}=V_{S} \\cdot \\left(\\dfrac{R_{P}+\\Delta R}{2 \\cdot R_{P}+\\Delta R}\\right) = ${V1_Sup} V\\)`,
    "eq-4": `\\(V''_{1}=V_{S} \\cdot \\left(\\dfrac{R_{P}-\\Delta R}{R_{P}-\\Delta R+R_{P}}\\right)\\)`,
    "eq-5": `\\(V''_{1}=V_{S} \\cdot \\left(\\dfrac{R_{P}-\\Delta R}{2 \\cdot R_{P}-\\Delta R}\\right) = ${V1_Inf} V\\)`,
    "eq-6": `\\(Ganho = \\dfrac{R_{B}}{R_{A}} = \\dfrac{${RB} K}{${RA} K} = ${ganho}\\\)`,
    "eq-7": `\\(V'_{o}=\\dfrac{R_{B}}{R_{A}} \\cdot \\left(V_{2}-V'_{1}\\right)=-V_{S} \\cdot \\dfrac{R_{B}}{R_{A}} \\cdot \\left(\\dfrac{\\Delta R}{4 \\cdot R_{P}+2 \\cdot \\Delta R}\\right) = ${Vo_Min} V\\)`,
    "eq-8": `\\(V'_{o}\\approx -V_{S} \\cdot \\dfrac{R_{B}}{R_{A}} \\cdot \\left(\\dfrac{1}{4 \\cdot R_{P}}\\right) \\cdot \\Delta R = ${Vo_Min_Apx} V\\)`,
    "eq-9": `\\(V''_{o}=\\dfrac{R_{B}}{R_{A}} \\cdot \\left(V_{2}-V'_{1}\\right)=V_{S} \\cdot \\dfrac{R_{B}}{R_{A}} \\cdot \\left(\\dfrac{\\Delta R}{4 \\cdot R_{P}-2 \\cdot \\Delta R}\\right) = ${Vo_Max} V\\)`,
    "eq-10": `\\(V''_{o}\\approx V_{S} \\cdot \\dfrac{R_{B}}{R_{A}} \\cdot \\left(\\dfrac{1}{4 \\cdot R_{P}}\\right) \\cdot \\Delta R = ${Vo_Max_Apx} V\\)`,
    "eq-11": `\\(${Vo_Min_Apx} V\\)`,
    "eq-12": `\\(${Vo_Max_Apx} V\\)`,
 }

  for (const key in equations) {
    const elementId = `${prefixo}-${key}`;
    const element = document.getElementById(elementId);
    if (element) {
      element.innerHTML = equations[key];
      if (window.MathJax && MathJax.typesetPromise) {
        MathJax.typesetPromise([element]);
      }
    }
  }

  const data_parameters = {
    VS: VS.toFixed(2),
    RP: RP.toFixed(2),
    RA: RA.toFixed(2),
    RB: RB.toFixed(2),
    RSGPos: RSGPos.toFixed(2),
  };

  const data_config = {
    components: ["V_{S}", "R_{P}", "R_{A}", "R_{B}","R_{SG}"],
    units: ["V", "\\Omega", "K\\Omega", "K\\Omega", "\\Omega"],
  };

  // Envie como JSON string
  const data_parameters_json = JSON.stringify(data_parameters);
  const data_config_json = JSON.stringify(data_config);

  if (typeof init_generic === "function") {
    init_generic(prefixo, data_parameters_json, data_config_json);
  }
}


function ex_DAC_refresh_input() {

  const prefixo = "ex_DAC";

  const bit_on = parseFloat(document.getElementById("ex_DAC-Bit-ON").value);
  
  const toogleV1 = document.getElementById("ex_DAC_toogle_V1");
  const toogleV2 = document.getElementById("ex_DAC_toogle_V2");
  const toogleV3 = document.getElementById("ex_DAC_toogle_V3");
  const toogleV4 = document.getElementById("ex_DAC_toogle_V4");
  
  let bitV1 = 0;
  let bitV2 = 0;
  let bitV3 = 0;
  let bitV4 = 0;
  
  if (toogleV1.checked) bitV1 = bit_on;
  if (toogleV2.checked) bitV2 = bit_on;
  if (toogleV3.checked) bitV3 = bit_on;
  if (toogleV4.checked) bitV4 = bit_on;
  
  equations = {
    "label-V1": `\\(=${bitV1}V\\)`,
    "label-V2": `\\(=${bitV2}V\\)`,
    "label-V3": `\\(=${bitV3}V\\)`,
    "label-V4": `\\(=${bitV4}V\\)`,    
  }

  for (const key in equations) {
    const elementId = `${prefixo}-${key}`;
    const element = document.getElementById(elementId);
    if (element) {
      element.innerHTML = equations[key];
      if (window.MathJax && MathJax.typesetPromise) {
        MathJax.typesetPromise([element]);
      }
    }
  }
}

function ex_DAC(action, mode = null) {
  const prefixo = ex_DAC.name;

  if (action == "reset") {
    reset(prefixo);
  }
  
  const bit_on = parseFloat(document.getElementById("ex_DAC-Bit-ON").value);
  
  const toogleV1 = document.getElementById("ex_DAC_toogle_V1");
  const toogleV2 = document.getElementById("ex_DAC_toogle_V2");
  const toogleV3 = document.getElementById("ex_DAC_toogle_V3");
  const toogleV4 = document.getElementById("ex_DAC_toogle_V4");
  
  let bitV1 = 0;
  let bitV2 = 0;
  let bitV3 = 0;
  let bitV4 = 0;
  
  if (toogleV1.checked) bitV1 = bit_on;
  if (toogleV2.checked) bitV2 = bit_on;
  if (toogleV3.checked) bitV3 = bit_on;
  if (toogleV4.checked) bitV4 = bit_on;

  const R1_mode1 = parseFloat(document.getElementById(`${prefixo}-R1-mode1`).value);
  const RF_mode1 = parseFloat(document.getElementById(`${prefixo}-RF-mode1`).value);

  const Vo_mode2 = parseFloat(document.getElementById(`${prefixo}-Vo-mode2`).value);
  const R1_mode2 = parseFloat(document.getElementById(`${prefixo}-R1-mode2`).value);

  let R1 = R1_mode1;
  let RF = RF_mode1;

  if(mode == 2){
    R1 = R1_mode2;
  }
  
  const R2 = R1*2;
  const R3 = R1*4;
  const R4 = R1*8;

  if(mode == 2){
    RF = Vo_mode2/((bit_on/R1)+(bit_on/R2)+(bit_on/R3)+(bit_on/R4));
  }

  const Vo = RF*((bitV1/R1)+(bitV2/R2)+(bitV3/R3)+(bitV4/R4));
  
  const data_parameters = {
    V1: bitV1.toFixed(2),
    V2: bitV2.toFixed(2),
    V3: bitV3.toFixed(2),
    V4: bitV4.toFixed(2),    
    R1: R1.toFixed(2),
    R2: R2.toFixed(2),
    R3: R3.toFixed(2),
    R4: R4.toFixed(2), 
    RF: RF.toFixed(2),    
    Vo: Vo.toFixed(4),           
  };

  const data_config = {
    components: ["V_{1}", "V_{2}", "V_{3}", "V_{4}","R_{1}", "R_{2}", "R_{3}", "R_{4}", "R_{F}", "-V_{o}"],
    units: ["V","V","V","V", "K\\Omega", "K\\Omega", "K\\Omega", "K\\Omega", "K\\Omega", "V"],
  };

  // Envie como JSON string
  const data_parameters_json = JSON.stringify(data_parameters);
  const data_config_json = JSON.stringify(data_config);

  if (typeof init_generic === "function") {
    init_generic(prefixo, data_parameters_json, data_config_json);
  }


}

File: aulas/aula_ampop_1/main.py
import js
import numpy as np # type: ignore
#import matplotlib.pyplot as plt

import sympy as sp # type: ignore
import cmath

import plotly.graph_objects as go # type: ignore

import json

import warnings
warnings.filterwarnings("ignore")

from pyscript import display # type: ignore

js.document.getElementById("pyscript-ready").setAttribute("ready", "true")


 
def ampop(e):     
    
    # Definir os parâmetros
    MaisVcc = 10
    MenosVcc = -15

    VN_cos_amplitude = 5
    VN_cos_freq = 10
    VN_cos_phase = 0
    VP_DC = 2
    A = float(js.document.getElementById("A_Gain").value)

    #A = 10  # Fator de amplificação

    # Calcular os limites de saturação
    MaisVcc_A = MaisVcc / A
    MenosVcc_A = MenosVcc / A

    # Gerar dados para o gráfico com 1000 pontos, limitando x até 2
    x = np.linspace(0, 1.2, 1000)  # Limitar o eixo x até 2
    y_cos = VN_cos_amplitude * np.cos(VN_cos_freq * x + VN_cos_phase)  # Senoide

    # Calcular a terceira onda Vo = A * (VP - VN) sem saturação
    Vo_without_saturation = A * (VP_DC - y_cos)

    # Aplicar saturação para Vo (limites de MaisVcc e MenosVcc)
    Vo = np.clip(Vo_without_saturation, MenosVcc, MaisVcc)

    # Criar figura
    fig = go.Figure(data=go.Scatter(x=x, y=y_cos, mode='lines', name='VN'))

    # Adicionar a linha de VP_DC
    fig.add_trace(go.Scatter(
        x=[x[0], x[-1]],  # Posições no eixo x
        y=[VP_DC, VP_DC],  # Posições no eixo y (valor constante VP_DC)
        mode='lines',  # Tipo de gráfico (linha)
        name='VP',  # Nome da linha
        line=dict(color='red')  # Linha contínua vermelha
    ))

    # Adicionar a terceira onda Vo (com saturação)
    fig.add_trace(go.Scatter(
        x=x, 
        y=Vo, 
        mode='lines', 
        name='Vo', 
        line=dict(color='green')  # Linha contínua verde
    ))

    # Adicionar linha vertical interativa para Vo sem saturação
    fig.update_layout(
        xaxis_title='tempo(s)',  # Troca de 'x' por 'tempo(s)'
        yaxis_title='Tensão(V)',  # Troca de 'Valores' por 'Tensão(V)'
        xaxis=dict(range=[0, 1.2]),  # Limitar o eixo x até 2
        hovermode='closest',  # Ativar hover para o ponto mais próximo
        showlegend=True,  # Exibe a legenda
    )

    # Função para atualizar a linha vertical ao mover o cursor
    fig.update_layout(
        shapes=[dict(
            type='line',
            x0=0,  # Começo da linha no eixo x
            x1=0,  # Fim da linha no eixo x (a linha será vertical)
            y0=MenosVcc,  # Começo da linha no eixo y
            y1=MaisVcc,  # Fim da linha no eixo y
            line=dict(color='black', dash='dot')  # Linha preta pontilhada
        )]
    )

    # Adicionar valores de Vo sem saturação no customdata
    fig.update_traces(
        customdata=np.vstack([Vo_without_saturation, y_cos]).T,  # Empilha Vo sem saturação e senoide para o customdata
        hovertemplate=(
            '<b style="font-size: 24px;">Tempo: %{x:.2f}s</b><br>' +
            '<span style="font-size: 24px;">A = ' + str(A) + '</span><br>' + 
            '<span style="font-size: 24px;">VN = %{customdata[1]:.2f}V</span><br>' +  # Agora, o valor de VN é a senoide na posição x
            '<span style="font-size: 24px;">VP: ' + str(VP_DC) + 'V</span><br>' +  # Valor fixo de VP_DC
            '<span style="font-size: 24px;">Vo = A(VP-VN) (sem saturação): %{customdata[0]:.2f}V</span><br>' +  # Valor Vo (sem saturação)
            '<span style="font-size: 24px;">Vo = A(VP-VN) (com saturação): %{y:.2f}V</span><br>'  # Valor Vo (com saturação)
        )
    )

    # Ajustar o layout do gráfico para ter mais espaço
    fig.update_layout(
        hoverlabel=dict(
                font_size=24,
                bgcolor='rgba(255, 255, 255, 0.5)'  # Fundo branco com 50% de transparência
            ),  # Ajustar o tamanho da fonte do hover        
        font=dict(size=24),
        margin=dict(l=20, r=20, t=50, b=20)  # Margens para evitar corte
    )

    display(fig, target="goComparador", append=False)



# Convertido para JS
def ex_AmpOp(e):
    
    prefixo = "ex_AmpOp"

    # VA = float(js.document.getElementById("ex_AmpOp-VA").value)
    # VB = float(js.document.getElementById("ex_AmpOp-VB").value)
    
    # MaisVcc = float(js.document.getElementById("ex_AmpOp-MaisVcc").value)
    # MenosVcc = float(js.document.getElementById("ex_AmpOp-MenosVcc").value)
    
    VA = float(js.document.getElementById(prefixo+"-"+"VA").value)
    VB = float(js.document.getElementById(prefixo+"-"+"VB").value)
    
    MaisVcc = float(js.document.getElementById(prefixo+"-"+"MaisVCC").value)
    MenosVcc = float(js.document.getElementById(prefixo+"-"+"MenosVCC").value)


    RS = 25*1e3
    RF = 100*1e3

    Vo = VB - ((VA-VB)/(RS))*RF

    Vo_str = repr(round(Vo, 2))

    equations = {
        "eq-1": "\(V_{n}=V_{p}=AAA\)",
        "eq-2": "\(i_{n}=i_{p}=0\)",
        "eq-3": "\(V_{p}=V_{B}\)",
        "eq-4": "\(V_{n}=V_{p}=V_{B}\)",
        "eq-5": "\(i_{S}=\\dfrac{V_{A}-V_{n}}{R_{s}}=\\dfrac{V_{A}-V_{B}}{R_{S}}\)",
        "eq-6": "\(i_{F}=\\dfrac{V_{n}-V_{o}}{R_{F}}=\\dfrac{V_{B}-V_{o}}{R_{F}}\)",
        "eq-7": "\(i_{S}=i_{F}+i_{n}\)",
        "eq-8": "\(i_{S}=i_{F}+0\)",
        "eq-9": "\(\\dfrac{V_{A}-V_{B}}{R_{S}}=\\dfrac{V_{B}-V_{o}}{R_{F}}\)",
        "eq-10": "\(R_{F} \\cdot \\left(\\dfrac{V_{A}-V_{B}}{R_{S}}\\right)-V_{B}=-V_{o}\)",
        "eq-11": "\(V_{o}=V_{B}-R_{F} \\cdot \\left(\\dfrac{V_{A}-V_{B}}{R_{S}}\\right)\)",
        "eq-12": "\(V_{o}="+str(VB)+"-100 \\cdot 10^{3} \\cdot \\left(\\dfrac{"+str(VA)+"-"+str(VB)+"}{25 \\cdot 10^{3}}\\right)="+str(Vo_str)+"V\)",
        "eq-13": "\(\\begin{cases} as=\\dfrac{1}{3} \\\\[0.1pt]  \\\\ as=\\dfrac{1}{3} \\end{cases}\)",

    }


    for key, equation in equations.items():
        element_id = f"{prefixo}-{key}"
        element = js.document.getElementById(element_id)
        if element:
            element.innerHTML = equation
            js.MathJax.Hub.Typeset(js.document.getElementById(element_id))



    # Dados como dicionários Python
    data_parameters = {'VA': repr(round(VA,2)), 'VB': repr(round(VB,2)), 'MaisVcc': repr(round(MaisVcc,2)), 'MenosVcc': repr(round(MenosVcc,2))}
    data_config = {
        'components': ['V_{A}', 'V_{B}', '+V_{cc}', '-V_{cc}'],
        'units': ['V', 'V', 'V', 'V']
    }

    # Converte os dicionários para objetos JS
    data_parameters_json = json.dumps(data_parameters)
    data_config_json = json.dumps(data_config)

    # Chama a função init_generic definida em JavaScript
    js.init_generic(prefixo, data_parameters_json, data_config_json)


ampop("")
ex_AmpOp("")
